lib/class-wp-rest-template-parts-controller.php

```diff
@@ -0,0 +1,130 @@
+<?php
+/**
+ * REST API: WP_REST_Template_Parts_Controller class
+ *
+ * @subpackage REST_API
+ * @package    WordPress
+ */
+
+/**
+ * Creates a template part auto-draft if it doesn't exist yet.
+ *
+ * @access private
+ *
+ * @param string $slug    Template part slug.
+ * @param string $theme   Template part theme.
+ * @param string $content Template part content.
+ */
+function create_auto_draft_for_template_part_block( $slug, $theme, $content ) {
+	// A published post might already exist if this template part
+	// was customized elsewhere or if it's part of a customized
+	// template. We also check if an auto-draft was already created
+	// because preloading can make this run twice, so, different code
+	// paths can end up with different posts for the same template part.
+	// E.g. The server could send back post ID 1 to the client, preload,
+	// and create another auto-draft. So, if the client tries to resolve the
+	// post ID from the slug and theme, it won't match with what the server sent.
+	$template_part_query = new WP_Query(
+		array(
+			'post_type'      => 'wp_template_part',
+			'post_status'    => array( 'publish', 'auto-draft' ),
+			'title'          => $slug,
+			'meta_key'       => 'theme',
+			'meta_value'     => $theme,
+			'posts_per_page' => 1,
+			'no_found_rows'  => true,
+		)
+	);
+	$template_part_post  = $template_part_query->have_posts() ? $template_part_query->next_post() : null;
+	if ( ! $template_part_post ) {
+		wp_insert_post(
+			array(
+				'post_content' => $content,
+				'post_title'   => $slug,
+				'post_status'  => 'auto-draft',
+				'post_type'    => 'wp_template_part',
+				'post_name'    => $slug,
+				'meta_input'   => array(
+					'theme' => $theme,
+				),
+			)
+		);
+	} else {
+		// Potentially we could decide to update the content if different.
```

Comment:
If we aren't comparing content as in the previous implementation, @david-szabo97 has been looking into re-writing the auto-drafts if the theme version changes - https://github.com/WordPress/gutenberg/pull/26383

Reply:
yes, with this simpler syncing function, it's just a matter of comparing content here and updating the auto-draft if needed.

Reply:
I think the updating of the auto-draft was already being done prior to #26383. The purpose of #26383 was to stop that code from running whenever it was not necessary to run it (i.e. theme version hasn't changed.) This PR seems to go backwards in that respect.

Reply:
It's because `gutenberg_resolve_template` calls `file_get_contents` on every template in the theme while `gutenberg_find_template_post_and_parts` only called it on the one being requested. That means if you are loading 10 templates in the site editor then `file_get_contents` will get called 100 times.

Reply:
Where do you see the template part update was being done before? By my own reading of the code, it was not being done anywhere.

Reply:
I was talking about templates actually. They were being updated inside `gutenberg_find_template_post_and_parts` if the file contents had changed. I don't know about template parts.

Reply:
> Where do you see the template part update was being done before? 

From what I can tell, the following was the closest to this.  It looks like it didn't so much update the auto-draft itself, but create a new one.  The current code only gets to this point if a `publish` version isn't found, but if the contents of the auto-draft don't match the contents of the file then a new auto-draft is created:

https://github.com/WordPress/gutenberg/blob/422b21c70ad6a28f1b1b2cd6acb642c3360fa454/lib/template-loader.php#L174-L178

Reply:
Yes, I can see that difference and it's very problematic IMO. Creating a new auto-draft when the post_content changes is probably not the best approach to synchronization. 

Let's leave this syncing behavior to its own dedicated PR since it's broken in master anyway (proliferation of auto-drafts).

Reply:
> it's very problematic IMO.
> Let's leave this syncing behavior to its own dedicated PR

Agreed, and that sounds like a good plan!